securedrop/journalist.py

```diff
@@ -430,15 +598,85 @@ def index():
             Submission.query.filter_by(source_id=source.id,
                                        downloaded=False).all())
 
-    return render_template('index.html', unstarred=unstarred, starred=starred)
+    return starred, unstarred, source_labels, all_source_labels
+
+
+@app.route('/')
+@login_required
+def index():
+    # Long SQLAlchemy statements look best when formatted according to
+    # the Pocoo style guide, IMHO:
+    # http://www.pocoo.org/internal/styleguide/
+    sources = Source.query.filter_by(pending=False) \
+                          .order_by(Source.last_updated.desc()) \
+                          .all()
+
+    starred, unstarred, labels, all_labels = get_stars_and_labels(sources)
+
+    return render_template('index.html', unstarred=unstarred, starred=starred,
+                           source_labels=labels, all_labels=all_labels)
+
+
+@app.route('/filter_by_label/<int:label_id>', methods=('POST',))
```

Comment:
In general, I like to think of views as having a 1-to-1-to-1 correspondence between the view's URL, view function, and template file. I think it makes the behavior of the web application more consistent and easier to modify.

Filtering a view is a classic case where GET parameters would be a good move. I would encourage you to:

1. Merge `filter_index` into `index` so the index view has a single view function
2. Supporting filtering via GET params, e.g. `?filter=foo,bar`. You'll have to take care to handle multiple filter arguments, and make sure you're url encoding/decoding properly.